// open-sml/lib/Constants.js
/*!
 * OpenSML
 * Copyright(c) 2014-2015 D. Spautz (d.spautz@web.de)
 * MIT Licensed
 */

const MAX_BUFFER_SIZE=512;

const REQUEST_FILE = 0x01;
const RESPONSE_FILE = 0x02;
const COMBI_FILE = 0x03;

const PUBLIC_OPEN_REQUEST=0x00000100;
const PUBLIC_OPEN_RESPONSE=0x00000101;
const PUBLIC_CLOSE_REQUEST=0x00000200;
const PUBLIC_CLOSE_RESPONSE=0x00000201;
const GET_PROFILE_PACK_REQUEST=0x00000300;
const GET_PROFILE_PACK_RESPONSE=0x00000301;
const GET_PROFILE_LIST_REQUEST=0x00000400;
const GET_PROFILE_LIST_RESPONSE=0x00000401;
const GET_PROC_PARAMETER_REQUEST=0x00000500;
const GET_PROC_PARAMETER_RESPONSE=0x00000501;
const SET_PROC_PARAMETER_REQUEST=0x00000600;
const SET_PROC_PARAMETER_RESPONSE=0x00000601;
const GET_LIST_REQUEST=0x00000700;
const GET_LIST_RESPONSE=0x00000701;
const ATTENTION_RESPONSE=0x0000FF01;

const UNSIGNED = 0x06;
const INTEGER = 0x05;
const BOOLEAN = 0x04;
const OCTET_STRING = 0x00;
const UINT8 = 0x62;
const UINT16 = 0x63;
const UINT32 = 0x65;
const UINT64 = 0x69;
const INT8 = 0x52;
const INT16 = 0x53;
const INT32 = 0x55;
const INT64 = 0x59;
const SML_VALUE = 0x01;
const SML_PERIOD_ENTRY = 0x02;
const SML_TUPEL_ENTRY = 0x03;
const SML_TIME = 0x04;
const TIMESTAMP = 0x02;
const SEC_INDEX = 0x01;
const OPTIONAL = 0x01;

module.exports = {
	MAX_BUFFER_SIZE: MAX_BUFFER_SIZE,
	PUBLIC_OPEN_REQUEST: PUBLIC_OPEN_REQUEST,
	PUBLIC_OPEN_RESPONSE: PUBLIC_OPEN_RESPONSE,
	PUBLIC_CLOSE_REQUEST: PUBLIC_CLOSE_REQUEST,
	PUBLIC_CLOSE_RESPONSE: PUBLIC_CLOSE_RESPONSE,
	GET_PROFILE_PACK_REQUEST: GET_PROFILE_PACK_REQUEST,
	GET_PROFILE_PACK_RESPONSE: GET_PROFILE_PACK_RESPONSE,
	GET_PROFILE_LIST_REQUEST: GET_PROFILE_LIST_REQUEST,
	GET_PROFILE_LIST_RESPONSE: GET_PROFILE_LIST_RESPONSE,
	GET_PROC_PARAMETER_REQUEST: GET_PROC_PARAMETER_REQUEST,
	GET_PROC_PARAMETER_RESPONSE: GET_PROC_PARAMETER_RESPONSE,
	SET_PROC_PARAMETER_REQUEST: SET_PROC_PARAMETER_REQUEST,
	SET_PROC_PARAMETER_RESPONSE: SET_PROC_PARAMETER_RESPONSE,
	GET_LIST_REQUEST: GET_LIST_REQUEST,
	GET_LIST_RESPONSE: GET_LIST_RESPONSE,
	ATTENTION_RESPONSE: ATTENTION_RESPONSE,
	UNSIGNED: UNSIGNED,
	INTEGER: INTEGER,
	BOOLEAN: BOOLEAN,
	OCTET_STRING: OCTET_STRING,
	TIMESTAMP: TIMESTAMP,
	SEC_INDEX: SEC_INDEX,
	UINT8: UINT8,
	UINT16: UINT16,
	UINT32: UINT32,
	UINT64: UINT64,
	INT8: INT8,
	INT16: INT16,
	INT32: INT32,
	INT64: INT64,
	SML_VALUE: SML_VALUE,
    SML_PERIOD_ENTRY: SML_PERIOD_ENTRY,
    SML_TUPEL_ENTRY: SML_TUPEL_ENTRY,
    SML_TIME: SML_TIME,
    OPTIONAL: OPTIONAL
};
